### ISSUE TITLE
Inconsistency between `setlocal` and `vim.opt_local` for `signcolumn`

### ISSUE THREAD

**Issue by krady21** [neovim/neovim#14670]
### TLDR
`vim.opt_local.signcolumn` seems to behave as window local
`setlocal signcolumn` seems to behave as buffer local

- `nvim --version`:
```
NVIM v0.5.0-dev
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```
- Operating system/version: Ubuntu 18.04.5 LTS
- Terminal name/version: GNOME Shell 3.28.4
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u mini_init.lua`

```
mkdir ~/.config/nvim/pack/test/start
git clone https://github.com/neovim/nvim-lspconfig.git ~/.config/nvim/pack/test/start/nvim-lspconfig
```

mini_init.lua:
```
local on_attach = function(client, bufnr)
     vim.opt_local.signcolumn = 'yes'
end
require'lspconfig'.rust_analyzer.setup{ on_attach = on_attach } -- or any lsp server for that matter
```
```
:e file.rs
:e Makefile
```
### Actual behaviour
Neovim sets signcolumn to 'yes'  in all the buffers that are opened afterwards in the window where the lsp attach function was first called, even if those buffers are not associated to any lsp server.

### Expected behaviour
Set `signcolumn` to `'yes'` only in buffers where lsp is used. Using `vim.cmd('setlocal signcolumn=yes')`in the `on_attach` function produces the desired behavior and makes the `signcolumn` setting buffer local (only applied to lsp attached buffers). This inconsistency probably comes from the fact the `signcolumn` is exposed as a window local option in lua, but I was hoping that adding opt_local in #13479  would fix it.

**Comment by mnarrell** [comment]
This behavior seems to apply to `spell` as well.  I've done `vim.opt_local.spell = true` for the markdown filetype.  After opening a markdown file, each subsequent buffer has spell checking enabled.

**Comment by tjdevries** [comment]
Hmmm, this is how window local options usually work. They are inherited from the window that you were in before, so when you split and do other things, I thought they should keep the same value.

For example, `signcolumn` is a window option, as mentioned in the help:

```
						*'signcolumn'* *'scl'*
'signcolumn' 'scl'	string	(default "auto")
			local to window
``

cc @bfredl am I crazy :)

**Comment by tjdevries** [comment]
Oops, accidentally tagged @bfredl in the code block :laughing:

**Comment by ii14** [comment]
@tjdevries If you do `:setl scl=yes`, then open a new buffer, `scl` on the new buffer is back to default value. It's basically as if it was local to buffer. See #14620

**Comment by tjdevries** [comment]
How are you opening a new buffer?  Via a split or via edit or what? Is the original buffer a listed, non-scratch buffer?

**Comment by ii14** [comment]
Just `:enew`. Previous buffer has to be modified to open a new one of course.

**Comment by francisco285** [comment]
@tjdevries Try this:
1. Use `vim.wo.signcolumn = 'yes'` in the on_attach function
2. Open neovim
3. `:e init.lua` (assuming this file will use lsp)
4. It will set signcolumn to yes on the current window
5. Using `:new` the new window also have signcolumn set to yes

Using `vim.cmd([[setlocal signcolumn=yes]])` in `1` results in `5` having signcolumn set to no (the default)

**Comment by ii14** [comment]
I don't really know what I'm talking about so this might make no sense, but could it be that vim docs are just misleading when they say that these options are local to window? Usually you don't want methods like `signcolumn`, `foldmethod` etc to be local to window. You want them to be local to buffer.

**Comment by tjdevries** [comment]
OK, Yeah, this option is confusing me :laughing: I'll have a chat w/ bfredl about it. Perhaps it's an issue with `nvim_win_set_option` (which is what `vim.opt`, and all the other ones are using under the hood).

It *is* a window option. That is how it's defined in the code. I don't know if there is special casing for this or if other things come into play. I'll have to look into that a bit more.

**Comment by bfredl** [comment]
The API is simply just incomplete for window options. Extending it properly will be a goal for 0.6.

**Comment by Shatur** [comment]
Also noticed the same issue when transferred my `ftplugin` settings that use `setlocal` to lua. `vim.wo` doesn't behave like `setlocal`.

**Comment by zeertzjq** [comment]
This issue also seems to exist for `conceallevel`.

**Comment by andrewferrier** [comment]
Curiously, `number` is local to window according to the documentation. So maybe vim's original implementation of `setlocal` is wrong, as it definitely appears to be local to a buffer (which makes a lot more sense to me).

**Comment by TheBlob42** [comment]
Since I just stumbled across this issue as well I did a bit of research and found `:h local-options`

```text
[...]
It's possible to set a local window option specifically for a type of buffer.
When you edit another buffer in the same window, you don't want to keep
using these local window options.  Therefore Vim keeps a global value of the
local window options, which is used when editing another buffer.  Each window
has its own copy of these values.  Thus these are local to the window, but
global to all buffers in the window.  With this you can do: >
	:e one
	:set list
	:e two
Now the 'list' option will also be set in "two", since with the ":set list"
command you have also set the global value. >
	:set nolist
	:e one
	:setlocal list
	:e two
Now the 'list' option is not set, because ":set nolist" resets the global
value, ":setlocal list" only changes the local value and ":e two" gets the
global value.  Note that if you do this next: >
	:e one
You will get back the 'list' value as it was the last time you edited "one".
The options local to a window are remembered for each buffer.  This also
happens when the buffer is not loaded, but they are lost when the buffer is
wiped out |:bwipe|.
```

I guess this is the part which is not correctly reflected by `vim.opt_local` and would explain the difference regarding `setlocal`

**Comment by lewis6991** [comment]
I'm pretty sure this issue is caused entirely by #9110 which has:

```c
    const int opt_flags = (type == SREQ_WIN && !(flags & SOPT_GLOBAL))
                          ? 0 : (type == SREQ_GLOBAL)
                                ? OPT_GLOBAL : OPT_LOCAL;
```

When before it was just:

```c
    const int opt_flags = (type == SREQ_GLOBAL) ? OPT_GLOBAL : OPT_LOCAL;
```

It looks like this was a bad fix for some weird behaviour in `winhl`.

**Comment by lewis6991** [comment]
> This issue also seems to exist for number and relativenumber. Using vim.opt_local.number = false turns off number for all buffers opened in a window and is not buffer-local (which is not desirable, at least to me!). setlocal nonumber in a .vim file has the desired effect.

This doesn't seem to be true (anymore). `vim.opt_local.number = false` only disables number for the current buffer. I believe #5996 fixed this.

**Comment by lewis6991** [comment]
Ok let's close.

Note, this issue still exists for `vim.wo`, but this is by design (#9110).

**Comment by andrewferrier** [comment]
Agreed with @lewis6991, not sure when this changed, but at least since 0.7, this seems to be working as expected and `vim.opt_local.number = false` (as an example) works as I expected - buffer-local, not window-local.
